### PR TITLE
[ENH] Automatic reference/stacks ordering

### PR DESCRIPTION
**Pull request by pdedumast**

### CODE REVIEW

**Comment by sebastientourbier** [comment]
@pdedumast After thinking more about the way to reorder scans I think the best way would be to create a new interface which will be inserted right after the DataGrabber. It will takes as input an ordered list of stacks `stack_order`, a list of `input_images` (and `inputs_manual` brain masks if it applies) and will generate the appropriate list of `output_images` and `output_masks`. These outputs could then be connected to the rest of the workflow.

**Comment by pdedumast** [comment]
@sebastientourbier after performing multiple tests (as you can partially see), I understaood that the different interfaces indeed mess up the order of the stacks between inputs and outputs. I will add a quick sorting of the inputs of each module, as it is done [here](https://nipype.readthedocs.io/en/0.11.0/users/tutorial_103.html). 

Apart from that, we mentioned that all retrieved stacks by the DataGrabber are preprocessed even though they are not used in the reconstruction. They are even involved in the histogram normalization which does influence the reconstruction...
I can easily "clean" this to preprocess only useful stacks. However, I'm afraid this will mess up all the tests... Any idea to fix this easily?

**Comment by sebastientourbier** [comment]
@pdedumast Great! Yes the link was exactly what I was talking about at the last meeting. For the tests you would have to update the file `.circleci/config.yml`. You will have to comment the following lines that makes the check of output differences here:
https://github.com/Medical-Image-Analysis-Laboratory/mialsuperresolutiontoolkit/blob/c036eba62d8920945c2cdb31809c1742d2f0dbfe/.circleci/config.yml#L162-L163
and here:
https://github.com/Medical-Image-Analysis-Laboratory/mialsuperresolutiontoolkit/blob/c036eba62d8920945c2cdb31809c1742d2f0dbfe/.circleci/config.yml#L196-L197
Then if you added a new input argument to the parser you can also update the call to docker run and test-01 or test-02. Here for test-01 as example:
https://github.com/Medical-Image-Analysis-Laboratory/mialsuperresolutiontoolkit/blob/c036eba62d8920945c2cdb31809c1742d2f0dbfe/.circleci/config.yml#L138-L147 

You can also create a new test following test-01 and test-02 as examples but you do so, the codacy-coverage-report job should be updated with the new code coverage report created here:
https://github.com/Medical-Image-Analysis-Laboratory/mialsuperresolutiontoolkit/blob/c036eba62d8920945c2cdb31809c1742d2f0dbfe/.circleci/config.yml#L229

File: pymialsrtk/interfaces/utils.py [hunk 1]
```diff
@@ -49,3 +49,57 @@ def run(command, env=None, cwd=None):
...
+    for s in p_order:
+        id_and_files_ordered.append([[s, f[1]] for f in orig_order if ('run-' + str(s)) in f[1]][0])
+
+    # # Todo: this if statement could be ignored to remove extra series.
+    # if len(p_files) > len(p_order):
+    #     remainings = [s for s in orig_order if s[0] not in p_order]
+    #     remainings.sort()
+    #     id_and_files_ordered = id_and_files_ordered + remainings
+
+    return [i[1] for i in id_and_files_ordered]
```
Comment (sebastientourbier) [inline pymialsrtk/interfaces/utils.py:105]: Seems a new line should be added to this file.

**Review by sebastientourbier — CHANGES_REQUESTED** [review]
@pdedumast Had time to review your PR and check CircleCI. Very nice work 👏 
I just let a few comments for typos that should be corrected and then, I guess you are all good to merge 👍

**Comment by sebastientourbier** [comment]
@pdedumast I just merged PR #47 that integrates the script wrapping call to the Docker command with updated doc. 

I also managed to solve: 
1. Issue #45 for code coverage (seems it was due to running nipype with the MultiProc plugin) and 
2. Issue #48 for bad memory allocation error, i.e `corrupted size vs. prev_size: 0x000055d4457176f0 ***` by using `tcmalloc` instead of `malloc`. 

**Note** 
By fixing code coverage, I also disable Nipype to generate provenance PROVN reports; it was causing issues with non-existing user/group id inside the container. But by doing that, it reduced the number of output files and so, I am sorry to tell you but you will need to update the `test-XX_outputs.txt` again. 😬

**Comment by pdedumast** [comment]
@sebastientourbier seems like our milestone for pymialsrtk v2.0.0 is ready! ;)

**Review by sebastientourbier — APPROVED** [review]
Seems all good to me! 
Very nice work @pdedumast ! 
I like the refactoring and the way you output the json file for Super-Resolution. Indeed they might a space in BIDS to add this specification when it gonna get refined by practice.
You can merge 🥳

**Comment by sebastientourbier** [comment]
Then, I would take of making the release 👍

**Pull request merged**